### DISCUSSION TITLE
How to access ffmpeg `astats` filter metadata through libmpv

### DISCUSSION THREAD

**Discussion by dweymouth** [mpv-player/mpv#14464] [Q&A] [answered]
I develop a desktop music player using libmpv and I would like to make use of the metadata from the [`astats`](https://ffmpeg.org/ffmpeg-filters.html#astats-1) filter to create a peak meter. I have tried prepending `astats=metadata=1` to my `af` fliter list which is applied using `mpv_set_property` (I also have equalizer af filters and those are working properly). I then try to observe the peak property with `mpv_observe_property(mpv, 0 /*userdata*/, "af-metadata/lavfi.astats.Overall.Peak_level", MPV_FORMAT_DOUBLE)` but I am not getting any property change events. Observing a different property like "time_pos" is working.

I am not sure what I am doing wrong. Basically trying to do what was described in https://github.com/mpv-player/mpv/issues/2645 but unfortunately there's no sample code in that issue once the original reporter got it working.

**Comment by richardpl** [comment]
I do not think IIRC that lavfi metadata is converted to mpv metadata...

**Reply by dweymouth** [reply]
It seems like it should be available under `af-metadata` https://mpv.io/manual/stable/#command-interface-af-metadata/%3Cfilter-label%3E

**Reply by richardpl** [reply]
Is there way to list all metadata entries? Then you could see if its there.
Also may depend if you use -af lavfi=[astats] or just -af  astats....

**Comment by kasper93** [comment, marked as answer]
`mpv -af "astats=metadata=1"` 
```lua
mp.get_property_native("af-metadata/astats.00")
```

or with label

`mpv -af "@myfilter:astats=metadata=1"` 
```lua
mp.get_property_native("af-metadata/myfilter")
```

This will return table with the keys from filter metadata like `lavfi.astats.Overall.Peak_level`.

You can do the same with C API with nodes.

Querying values directly like `mp.get_property_native("af-metadata/astats.00/lavfi.astats.Overall.Peak_level")` is currently broken, after a05b847879dbb5d30bebf188562d16fa3e68ca70.

**Reply by kasper93** [reply]
> Querying values directly like `mp.get_property_native("af-metadata/astats.00/lavfi.astats.Overall.Peak_level")` is currently broken, after [a05b847](https://github.com/mpv-player/mpv/commit/a05b847879dbb5d30bebf188562d16fa3e68ca70).

Will be fixed by #14527

**Comment by dweymouth** [comment]
Thanks @kasper93 this worked great! In case this helps anyone else here's how I implemented this. My project is written in Go but I implemented this particular functionality in C to avoid the mpv_node having to cross the FFI boundary.

My `af` property string has `@astats:astats=metadata=1:reset=1:measure_overall=none` to enable the astats filter to compute the data and assign the `astats` property label.

https://github.com/dweymouth/supersonic/blob/92d2fb0/backend/player/mpv/peaks.c
